lib/deep_cover/autoload_tracker.rb

```diff
@@ -0,0 +1,212 @@
+
+module DeepCover
+  class AutoloadTracker
+    AutoloadEntry = Struct.new(:weak_mod, :name, :target_path, :interceptor_path) do
+      # If the ref is dead, will return nil
+      # If the target is frozen, will warn and return nil
+      # Otherwise return the target
+      def mod_if_available
+        mod = weak_mod.__getobj__
+        if mod.frozen?
+          AutoloadTracker.warn_frozen_module(mod)
+          nil
+        else
+          mod
+        end
+      rescue WeakRef::RefError
+        nil
+      end
+    end
+
+    attr_reader :autoloads_by_basename, :interceptor_files_by_path
+    def initialize
+      @autoloads_by_basename = {}
+      @interceptor_files_by_path = {}
+    end
+
+    def autoload_path_for(mod, name, path)
+      interceptor_path = setup_interceptor_for(mod, name, path)
+
+      if DeepCover.custom_requirer.is_being_required?(path)
+        already_loaded_feature
+      else
+        interceptor_path
+      end
+    end
+
+    def possible_autoload_target?(requested_path)
+      basename = basename_without_extension(requested_path)
+      autoloads = @autoloads_by_basename[basename]
+      autoloads && !autoloads.empty?
+    end
+
+    def wrap_require(requested_path, absolute_path_found, &block)
+      entries = entries_for_target(requested_path, absolute_path_found)
+
+      begin
+        entries.each do |entry|
+          mod = entry.mod_if_available
+          next unless mod
+          mod.autoload_without_deep_cover(entry.name, already_loaded_feature)
+        end
+
+        yield
+      ensure
+        entries = entries_for_target(requested_path, absolute_path_found)
+        entries.each do |entry|
+          mod = entry.mod_if_available
+          next unless mod
+          # Putting the autoloads back back since we couldn't complete the require
+          mod.autoload_without_deep_cover(entry.name, entry.interceptor_path)
+        end
+      end
+    end
+
+    # This is only used on MRI, so ObjectSpace is alright.
+    def initialize_autoloaded_paths(mods = ObjectSpace.each_object(Module), &do_autoload_block)
+      mods.each do |mod|
+        # Module's constants are shared with Object. But if you set autoloads directly on Module, they
+        # appear on multiple classes. So just skip, Object will take care of those.
+        next if mod == Module
+
+        if mod.frozen?
+          if mod.constants.any? { |name| mod.autoload?(name) }
+            self.class.warn_frozen_module(mod)
+          end
+          next
+        end
+
+        mod.constants.each do |name|
+          path = mod.autoload?(name)
+          next unless path
+          interceptor_path = setup_interceptor_for(mod, name, path)
+          yield mod, name, interceptor_path
+        end
+      end
+    end
+
+    # We need to remove the interceptor hooks, otherwise, the problem if manually requiring
+    # something that is autoloaded will cause issues.
+    def remove_interceptors(&do_autoload_block)
```

Comment:
Actually, that method does *not* `remove_interceptors`, it simply iterates on them...

Reply:
There is another review comment about this below with the discussion.